juju/juju
# [JUJU-1616] Add backend for secrets access control

This PR implements the secret ACL backend. The stub permission checks on the secrets manager facade were reworked to suit. It also fixes a few issues with the grant/revoke hook commands that appeared when the full QA was done.

## Checklist


## QA steps

deploy test charms
```
juju deploy juju-qa-dummy-source
juju deploy juju-qa-dummy-sink
juju add-unit dummy-sink
```
create a secret
```
juju exec --unit dummy-source/0 "secret-add data=foo"
secret:cbtcu5efg3o24llgfr7g
```
create check the owner can update and read a secret
```
juju exec --unit dummy-source/0 "secret-update secret:cbtcu5efg3o24llgfr7g data=foo2"
juju exec --unit dummy-source/0 "secret-get secret:cbtcu5efg3o24llgfr7g"
data: foo2
```
check that unauthorised access not allowed
```
juju exec --unit dummy-sink/0 "secret-get secret:cbtcu5efg3o24llgfr7g"
ERROR permission denied
```
relate dummy-sourc to dummy-sink and grant access
```
juju relate dummy-source dummy-sink
juju exec --unit dummy-source/0 "secret-grant secret:cbtcu5efg3o24llgfr7g -r 0"
```
any unit of dummy-sink can read the secret
```
juju exec --unit dummy-sink/0 "secret-get secret:cbtcu5efg3o24llgfr7g"
data: foo2
juju exec --unit dummy-sink/1 "secret-get secret:cbtcu5efg3o24llgfr7g"
data: foo2
```
remove the relation and access goes away
```
juju remove-relation dummy-source:sink dummy-sink:source
juju exec --unit dummy-sink/1 "secret-get secret:cbtcu5efg3o24llgfr7g"
ERROR permission denied
```
add relation again and grant access just to dummy-sink/1
```
juju relate dummy-source dummy-sink
juju exec --unit dummy-source/0 "secret-grant secret:cbtcu5efg3o24llgfr7g -r 1 --unit dummy-sink/1"
juju exec --unit dummy-sink/1 "secret-get secret:cbtcu5efg3o24llgfr7g"
data: foo2
juju exec --unit dummy-sink/0 "secret-get secret:cbtcu5efg3o24llgfr7g"
ERROR permission denied
```
revoke access and check
```
juju exec --unit dummy-source/0 "secret-revoke secret:cbtcu5efg3o24llgfr7g --unit dummy-sink/1"
uju exec --unit dummy-sink/1 "secret-get secret:cbtcu5efg3o24llgfr7g"
ERROR permission denied
```
